src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int ARRAY_START_POSITION = 0;
+    private Object[] listOfData;
+    private int size;
+
+    public ArrayList() {
+        listOfData = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkLength();
+        listOfData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        checkLength();
+        copy(index, index + 1);
+        listOfData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheck(index);
+        return (T) listOfData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheck(index);
+        listOfData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheck(index);
+        T value = (T)listOfData[index];
+        copy(index + 1, index);
+        size--;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getIndex(element);
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Object[] grow() {
+        Object[] temp = new Object[size + (size >> 1)];
+        copyToGrow(listOfData, temp);
+        return temp;
+    }
+
+    private void checkLength() {
+        if (size == listOfData.length) {
+            listOfData = grow();
+        }
+    }
```

Comment:
we can move grow() and copyToGrow() code into this method
there is no need to create private methods 2-3 lines long if they are not used in multiple places